Security_Engine/Compute/CameraFieldOfView.cs

```diff
@@ -0,0 +1,351 @@
+
+            //create and sort extended ray lines passing through intersected points
+            List<Line> rayLines = new List<Line>();
+            foreach (Point point in intersectPoints)
+            {
+                Line line = BH.Engine.Geometry.Create.Line(cameraLocation, point);
+                Line rayLine = line.Extend(0, radius - line.Length());
+                rayLines.Add(rayLine);
+            }
+            rayLines = rayLines.CullDuplicateLines();
+            rayLines = rayLines.OrderBy(x => x.SingedAngle(rayLines[0], Vector.ZAxis)).ToList();
+
+            //split ray lines and find visible line
+            List<Dictionary<Line, Polyline>> linesDict = new List<Dictionary<Line, Polyline>>();
+            foreach (Line rayLine in rayLines)
+            {
+                Line visibleLine = rayLine.VisibleLine(intersectObstacles);
+                linesDict.Add(visibleLine.LineObstacleDictionary(intersectObstacles));
+            }
+
+            //create points chain
+            List<Point> pointsChain = linesDict.PointsChain(cameraLocation, radius, distanceTolerance);
+
+            //create cone
+            Arc coneArc = cameraCone.Curves[1] as Arc;
+            PolyCurve cameraViewPolyCurve = pointsChain.CameraViewPolyCurve(cameraLocation, radius, coneArc, distanceTolerance);
+
+            return cameraViewPolyCurve;
+        }
+
+        /***************************************************/
+
+        [Description("Project obstacle polyline on camera plane.")]
+        [Input("obstacle", "Obstacle to project.")]
+        [Input("cameraPlane", "Plane to project the obstacle on.")]
+        [Output("newObstacle", "Obstacle polyline projected on camera plane.")]
+
+        private static Polyline FixAndProject(this Polyline obstacle, Plane cameraPlane, double tolerance)
+        {
+            if (obstacle.IsInPlane(cameraPlane))
+                return obstacle;
+
+            if (!obstacle.IsClosed(tolerance) || !obstacle.IsPlanar(tolerance))
+            {
+                Base.Compute.RecordWarning("Obstacle polyline is open and not planar. It will be closed and projected on camera plane.");
+                List<Point> projPoints = obstacle.ControlPoints.Select(x => x.Project(cameraPlane)).ToList();
+                return BH.Engine.Geometry.Create.Polyline(projPoints);
+            }
+            else if (obstacle.Normal().IsEqual(cameraPlane.Normal, tolerance) || obstacle.Normal().IsEqual(-cameraPlane.Normal, tolerance))
+            {
+                Base.Compute.RecordWarning("Obstacle polyline is parallel, but not in camera plane. It will be projected on camera plane.");
+                return obstacle.Project(cameraPlane);
+            }
+            else
+            {
+                Base.Compute.RecordWarning("Polyline is in different plane than camera plane. It will be projected on camera plane.");
+                return obstacle.Project(cameraPlane);
```

Comment:
I can see you do `return obstacle.Project(cameraPlane)` in every part of the if/else statement - it could be moved outside of the statement, after it

Reply:
Good point, I reorganized loop in that way 👍